### PR TITLE
Bluetooth: fix HCI command / data deadlock

### LINKED ISSUE #25917: Bluetooth: Deadlock with TX of ACL data and HCI commands (command blocked by data)
**Describe the bug**
Performing GATT write commands from a peripheral and a central as fast as possible from the main loop while the peer central is performing security procedure to bond, causes the following assertion fail on both the peripheral and central device

In peripheral:
```
*** Booting Zephyr OS build v2.3.0-rc1-368-g8fecd4b32d9d  ***
Bluetooth initialized
[00:00:00.007,507] <inf> fs_nvs: 6 Sectors of 4096 bytes
[00:00:00.007,507] <inf> fs_nvs: alloc wra: 0, ff0
[00:00:00.007,507] <inf> fs_nvs: data wra: 0, 0
[00:00:00.008,941] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
[00:00:00.008,941] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
[00:00:00.008,941] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 2.3 Build 0
[00:00:00.009,216] <inf> bt_hci_core: No ID address. App must call settings_load()
[00:00:00.031,829] <inf> bt_hci_core: Identity: c1:ab:66:13:11:2e (random)
[00:00:00.031,860] <inf> bt_hci_core: HCI: version 5.2 (0x0b) revision 0x0000, manufacturer 0x05f1
[00:00:00.031,860] <inf> bt_hci_core: LMP: version 5.2 (0x0b) subver 0xffff
Advertising successfully started
Connected: ea:15:23:d0:ac:03 (random)
MTU: 23
Exchange pending...Write: count= 0, len= 0, rate= 0 bps.
MTU exchange successful (247)
ASSERTION FAIL [err == 0] @ WEST_TOPDIR/zephyr/subsys/bluetooth/host/hci_core.c:370
        k_sem_take failed with err -11
[00:00:27.460,693] <err> os: r0/a1:  0x00000003  r1/a2:  0x00000001  r2/a3:  0x00000001
[00:00:27.460,693] <err> os: r3/a4:  0x00024c28 r12/ip:  0x00000000 r14/lr:  0x00007a01
[00:00:27.460,693] <err> os:  xpsr:  0x61000000
[00:00:27.460,723] <err> os: Faulting instruction address (r15/pc): 0x00007a0c
[00:00:27.460,723] <err> os: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
[00:00:27.460,723] <err> os: Current thread: 0x200014f0 (unknown)
[00:00:27.807,037] <err> os: Halting system
```
In central_hr:
```
*** Booting Zephyr OS build v2.3.0-rc1-368-g8fecd4b32d9d  ***
Bluetooth initialized
Scanning successfully started
[DEVICE]: 22:52:48:f8:a5:91 (random), AD evt type 3, AD data len 31, RSSI -45
[DEVICE]: 63:47:03:75:bc:8f (random), AD evt type 0, AD data len 29, RSSI -41
[AD]: 1 data_len 1
[AD]: 3 data_len 6
Connected: 63:47:03:75:bc:8f (random)
MTU: 23
Exchange pending...[00:00:00.006,042] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)MTU exchange successful (247)

[00:00:00.006,042] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
[00:00:00.006,042] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 2.3 Build 0
[00:00:00.006,683] <inf> bt_hci_core: Identity: ea:15:23:d0:ac:03 (random)
[00:00:00.006,683] <inf> bt_hci_core: HCI: version 5.2 (0x0b) revision 0x0000, manufacturer 0x05f1
[00:00:00.006,683] <inf> bt_hci_core: LMP: version 5.2 (0x0b) subver 0xffff
[00:00:00.688,507] <err> bt_conn: Unable to allocate TX context
ASSERTION FAIL [err == 0] @ WEST_TOPDIR/zephyr/subsys/bluetooth/host/hci_core.c:370
        k_sem_take failed with err -11
[00:00:11.095,764] <err> os: r0/a1:  0x00000003  r1/a2:  0x00000001  r2/a3:  0x00000001
[00:00:11.095,764] <err> os: r3/a4:  0x0002142a r12/ip:  0x00000000 r14/lr:  0x0000666d
[00:00:11.095,764] <err> os:  xpsr:  0x61000000
[00:00:11.095,764] <err> os: Faulting instruction address (r15/pc): 0x00006678
[00:00:11.095,794] <err> os: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
[00:00:11.095,794] <err> os: Current thread: 0x2000117c (unknown)
[00:00:11.702,239] <err> os: Halting system
```

**To Reproduce**
Use the branch: https://github.com/cvinayak/zephyr/commit/a2289565363a210e33b8956eac46df55e19a790a

Build and flash peripheral
1. mkdir -p build/peripheral; cd build/peripheral
2. cmake -GNinja -DBOARD=nrf52dk_nrf52832 ../../samples/bluetooth/peripheral
3. ninja
4. ninja flash

Build and flash central_hr
5. mkdir -p build/central_hr; cd build/central_hr
6. cmake -GNinja -DBOARD=nrf52dk_nrf52832 ../../samples/bluetooth/central_hr
7. ninja
8. ninja flash

9. Open two terminals, `minicom -D /dev/ttyACMx`
10. Observe both peripheral and central_hr fail after connection

**Expected behavior**
peripheral and central_hr should connect, transfer write commands, perform SMP pairing, be encrypted and continue to transder write commands.

**Impact**
showstopper

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK 0.11.2
 - Commit SHA or Version used: https://github.com/cvinayak/zephyr/commit/a2289565363a210e33b8956eac46df55e19a790a

### PR DESCRIPTION
**Pull request by jori-nordic**
See commits for more info. 
TLDR: makes the ACL fragmentation asynchronous, freeing the TX thread for sending commands when the ACL buffers are full.

Fixes #25917

### CODE REVIEW

File: subsys/bluetooth/host/conn.c [hunk 1]
```diff
@@ -44,6 +44,7 @@
 
 struct tx_meta {
 	struct bt_conn_tx *tx;
+	bool frag;
```
Comment (jori-nordic) [inline subsys/bluetooth/host/conn.c:47]: might rename that to `cont` if it's clearer

File: subsys/bluetooth/host/conn.c [hunk 2]
```diff
@@ -789,10 +852,13 @@ void bt_conn_process_tx(struct bt_conn *conn)
 		return;
 	}
 
-	/* Get next ACL packet for connection */
-	buf = net_buf_get(&conn->tx_queue, K_NO_WAIT);
+	/* Get next ACL packet for connection. The buffer will only get dequeued
+	 * if there is a free controller buffer to put it in.
+	 */
+	buf = k_fifo_peek_head(&conn->tx_queue);
```
Comment (jhedberg) [inline subsys/bluetooth/host/conn.c:858]: The problem with this is that you're skipping the fragment handling that `net_buf_get()` takes care of. We don't (AFAIK) currently take advantage of the net_buf fragment feature in the Bluetooth subsystem, but at least we've used the proper net_buf APIs until now that should keep the code working if we did use fragments.

At the very least, in addition to asserting that buf != NULL, you should also `BT_ASSERT((buf->flags & NET_BUF_FRAGS) == 0);`. Though even that's pretty unclean, since these are really net_buf-internal features.

A cleaner way would be to introduce a new `net_buf_peek()` API that does the same as `net_buf_get()` but also fetches any pending fragments.

Yet another, perhaps more long term solution, would be to finally move the net_buf fragments list into it's own independent list that's not mangled together with the linked list pointer as part of the same union (which is the reason for the NET_BUF_FRAGS flag and why fragments are placed as independent objects into the FIFO). Currently the net_buf definition starts off like this:
```c
struct net_buf {
	union {
		/** Allow placing the buffer into sys_slist_t */
		sys_snode_t node;

		/** Fragments associated with this buffer. */
		struct net_buf *frags;
	};
```

The penalty of moving `frags` out from the union would be that the struct size increases by the size of a pointer, but the benefit would be that we don't have this hassle of serializing & deserializing fragments whenever buffers pass through LIFO or FIFO objects.

File: subsys/bluetooth/host/conn.c [hunk 3]
```diff
@@ -567,35 +598,47 @@ static bool send_frag(struct bt_conn *conn, struct net_buf *buf, uint8_t flags,
...
+		uint16_t frag_len = MIN(conn_mtu(conn), net_buf_tailroom(frag));
+
+		net_buf_add_mem(frag, buf->data, frag_len);
+		net_buf_pull(buf, frag_len);
+	} else {
+		/* De-queue the buffer now that we know we can send it.
+		 * Only applies if the buffer to be sent is the original buffer,
+		 * and not one of its fragments.
+		 */
+		struct net_buf *dequeued = net_buf_get(&conn->tx_queue, K_NO_WAIT);
```
Comment (jhedberg) [inline subsys/bluetooth/host/conn.c:629]: The code comment should explicitly mention that the buffer was fetched from the FIFO using a "peek" operation.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/conn.c (hunk 2 above)
Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:858]: > At the very least, in addition to asserting that buf != NULL, you should also `BT_ASSERT((buf->flags & NET_BUF_FRAGS) == 0);`. Though even that's pretty unclean, since these are really net_buf-internal features.

Oh, and in addition to the assert, you need to explicitly set `buf->frags = NULL` since the k_fifo implementation leaves the linked list pointer untouched when removing from the list.

Also see https://github.com/zephyrproject-rtos/zephyr/pull/38830#pullrequestreview-763305021 and related discussion in the PR.

Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:858]: is that really a problem since I'm using the `_get` before passing it to `bt_send`? That takes care of the fragments, and the peek is a non-destructive operation afaik.

Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:858]: At the very least you need to clear the `buf->frags` pointer since `k_fifo_peek_head()` doesn't do that for you. You have the `EIO` branch below where e.g. `net_buf_unref()` is called, and `net_buf_unref()` will also unref the frags pointer if it is non-NULL. Same goes for any other net_buf API calls you might do on the buffer before calling `net_buf_get()`

Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:858]: Actually, I just realised that the `EIO` branch is probably buggy in this regard, i.e. of `send_buf()` returns error before getting to the `net_buf_get()` phase, you have a pointer to something that's still in the FIFO and you don't own a reference to, yet you still call `net_buf_unref()` on it.

Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:858]: This `frags` business is really backwards IMO, why do we even use it? seems like a linked-list implementation that we then put on another linked-list (the queue)?
Then we need all those safe APIs to deal with that instead of being able to just use the fifo/queue apis directly.
anyway for the EIO case, the buffer has already been consumed by the `_get`.

Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:858]: > Why do we want to clear the frags ptr again? isn't anything in frags actually part of the current buffer?

Yes, but only if those buffers were removed from the FIFO that they were passed through. If you use k_fifo APIs directly you may end up with a frags pointing at buffers that are still inside the FIFO, i.e. it's the FIFO that owns their references, not the parent buffer. Since we don't use frags (it's a feature added by the networking stack folks) we could for now just assert that frags were not used and then explicitly set the buf->frags pointer to NULL.

Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:858]: Anyway, as I said, really the cleanest way out of this would be to detach the frags from being in a union together with the linked list pointer in net_buf. But that's also considerably more work, more intrusive (affecting things outside of Bluetooth) and also comes with an increased memory footprint.

Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:858]: > This frags business is really backwards IMO, why do we even use it? seems like a linked-list implementation that we then put on another linked-list (the queue)?

You'll have to as the networking subsystem folks about that. I don't think they use FIFOs as much as we do, but then they do have an abstraction called net_pkt which has a pointer to the "top" buffer, which may then have a fragment list.

> anyway for the EIO case, the buffer has already been consumed by the _get.

Even then, this code looks pretty fragile, and could break if it's modified by someone not intimately familiar with net_buf internals. Anyway, if you make sure you always own a reference to the buffer when having a pointer to it (i.e. call net_buf_ref() after peek()) then the code becomes more future-proof.

Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:858]: I also thought we weren't using frags (ie I had an assert doing just that before), but there seem to be some use of it in `l2cap.c` regarding SDU handling.
I'm not sure I'm reading the code correctly though, so I'd appreciate if you took a look just to double-check.

Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:858]: One compromise I just thought of: maybe we could introduce a new CONFIG_NET_BUF_FRAGS option to enable/disable net_buf fragments support. Since we don't use/need it, we could keep it disabled, as long as you don't enable some other subsystem that depends on it (like the networking stack itself).

Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:858]: > introduce a new CONFIG_NET_BUF_FRAGS option 

Well, that'd take care of the memory overhead that'd come with `frags` being outside of the union, but we'd still need to use the proper APIs so that our code doesn't break if the option is enabled.

Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:858]: > Even then, this code looks pretty fragile, and could break if it's modified by someone not intimately familiar with net_buf internals.
I think this can be argued for the whole BT host. I know I've broken a bunch of stuff because of hidden assumptions like these.
I thought about moving the error handling somewhere else, but all the other places seemed messy as well. I'll at least add the referencing (and frags assert) as you said.

Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:858]: > I also thought we weren't using frags (ie I had an assert doing just that before), but there seem to be some use of it in l2cap.c regarding SDU handling.

Right, I had forgotten about that. IIRC that's code for 6LoWPAN/IPSP when we receive buffers from the Networking subsystem that we should then send over L2CAP Connection-oriented Channels. In such a case the Net subsystem might feed us with fragment lists.

Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:858]: It looks like the `l2cap.c` frags code stays within the L2CAP layer, i.e. when passing to ACL (conn->tx_queue) there are no fragment buffers anymore.

Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:858]: thanks for looking into it. that was my initial understanding as well, just wanted to be sure, since we pass the buffers around and I'm not always sure what we do with those frags.

I will add the assert when we put the buffers in the queue, I think it's safer.

File: subsys/bluetooth/host/conn.c [hunk 4]
```diff
@@ -497,24 +500,43 @@ static int send_iso(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
...
+	}
+#endif /* CONFIG_BT_ISO */
+#if defined(CONFIG_BT_CONN)
+	return bt_dev.le.acl_mtu;
+#else
+	return 0;
+#endif /* CONFIG_BT_CONN */
+}
+
+static int _send_frag(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
```
Comment (carlescufi) [inline subsys/bluetooth/host/conn.c:522]: This should be `static int do_send_frag()`

File: subsys/bluetooth/host/conn.c (hunk 2 above)
Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:858]: Can I have it at both places? because good luck getting a stack trace of who put that bad buffer in if it hits when popping it rather than when pushing it in.

Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:858]: Yes, that's fine. We could make the writing side a runtime check too, if we consider this the public API for that layer.

Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:857]: > Oh, and in addition to the assert, you need to explicitly set buf->frags = NULL since the k_fifo implementation leaves the linked list pointer untouched when removing from the list.

I just tried and it didn't work. I think it's because `buf->frags` is in a union with the node ptr. 
So if I overwrite it with `NULL` while it is still on the queue, and it ends up that I don't pull it from the fifo (e.g. the sem_take fails), the fifo list is effectively broken.

Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:857]: Right, good point. And this is another reason why using peak() together with the union is bad - you end up with the buffer in an inconsistent state until you actually remove it from from the FIFO.

Comment (jhedberg) [reply subsys/bluetooth/host/conn.c:857]: The short term solution would then probably be to document this clearly and try to make the code flow path from peek() to get() as short and obvious as possible.

Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:857]: yep agreed. I will place a comment above the `peek` saying this.

Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:860]: so I reverted the assert since #52760 got merged.

**Comment by jori-nordic** [comment]
@jhedberg could you review again?

**Comment by carlescufi** [comment]
@runegrunnet could you please take a look?

**Review by jhedberg — DISMISSED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/prj.conf [hunk 5]
```diff
@@ -18,7 +18,14 @@ CONFIG_BT_GAP_AUTO_UPDATE_CONN_PARAMS=n
 # L2CAP MPS
 # 23+27+27=77 makes exactly three full packets
 CONFIG_BT_L2CAP_TX_MTU=77
-CONFIG_BT_BUF_ACL_TX_SIZE=77
+
+# Use this to send L2CAP PDUs without any fragmentation.
+# In this particular case, we prefer fragmenting to test that code path.
+# CONFIG_BT_BUF_ACL_TX_SIZE=81
```
Comment (Thalley) [inline tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/prj.conf:24]: Why is this changed from 77 to 81? Header size not taken into account before?

File: tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/src/main.c [hunk 6]
```diff
@@ -267,6 +276,11 @@ static void test_peripheral_main(void)
 	LOG_DBG("*L2CAP STRESS Peripheral started*");
 	int err;
 
+	/* Prepare tx_data */
+	for (int i = 0; i < sizeof(tx_data); i++) {
```
Comment (Thalley) [inline tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/src/main.c:280]:
```suggestion
	for (size_t i = 0; i < sizeof(tx_data); i++) {
```

**Review by Thalley — CHANGES_REQUESTED** [review]

**Comment by jori-nordic** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/37577 was very useful for reproducing the deadlock in simulation, so thanks a lot @prathje ! You might want to review this as well, or try it on hardware.

File: tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/src/main.c [hunk 7]
```diff
@@ -267,6 +276,11 @@ static void test_peripheral_main(void)
...
+		tx_data[i] = i % 256;
```
[the remark this answers is not in the document]
Comment (jori-nordic) [reply tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/src/main.c:281]: if the compiler isn't doing this automatically (modulo power of two), then it's not a very good compiler :p
I'll change it, I don't really care

File: tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/prj.conf (hunk 5 above)
Comment (jori-nordic) [reply tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/prj.conf:24]: yep

File: subsys/bluetooth/host/conn.c [hunk 8]
```diff
@@ -619,52 +660,56 @@ static struct net_buf *create_frag(struct bt_conn *conn, struct net_buf *buf)
...
-			return false;
+			return -ENOMEM;
 		}
 
-		if (!send_frag(conn, frag, FRAG_CONT, true)) {
-			return false;
+		err = send_frag(conn, buf, frag, flags);
+		if (err) {
+			LOG_DBG("%p failed, mark as existing frag", buf);
+			tx_data(buf)->frag = flags != FRAG_START;
```
[the remark this answers is not in the document]
Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:702]: yeah I agree. The naming of this variable could be better.
This mark/tag is only used when this function can exit without having sent the packet. 
But in this case, a start fragment _will_ be noted as a fragment. This loop is used for the start & cont frags. The last frag is sent after exiting the loop.

File: subsys/bluetooth/host/conn.c [hunk 9]
```diff
@@ -567,35 +598,45 @@ static bool send_frag(struct bt_conn *conn, struct net_buf *buf, uint8_t flags,
...
+		LOG_DBG("no controller bufs");
+		return -ENOBUFS;
 	}
-#endif /* CONFIG_BT_BREDR */
-#if defined(CONFIG_BT_ISO)
-	if (conn->type == BT_CONN_TYPE_ISO && bt_dev.le.iso_mtu) {
-		return bt_dev.le.iso_mtu;
+
+	/* Add the data to the buffer */
+	if (frag) {
```
[the remark this answers is not in the document]
Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:619]: yes. I will probably rename it.

File: subsys/bluetooth/host/conn.c [hunk 10]
```diff
@@ -497,24 +500,43 @@ static int send_iso(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
...
 	struct bt_conn_tx *tx = tx_data(buf)->tx;
 	uint32_t *pending_no_cb;
 	unsigned int key;
 	int err = 0;
 
-	LOG_DBG("conn %p buf %p len %u flags 0x%02x", conn, buf, buf->len, flags);
-
-	/* Wait until the controller can accept ACL packets */
-	k_sem_take(bt_conn_get_pkts(conn), K_FOREVER);
+	LOG_HEXDUMP_DBG(buf->data, buf->len, "send_frag");
```
[the remark this answers is not in the document]
Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:529]: yeah you're right. I hadn't considered a case where logs are stored somewhere instead of just printed out over some comm interface.
I think I'll just drop those, one can easily add them back for debugging.

File: subsys/bluetooth/host/conn.c [hunk 11]
```diff
@@ -567,35 +598,45 @@ static bool send_frag(struct bt_conn *conn, struct net_buf *buf, uint8_t flags,
...
 {
-#if defined(CONFIG_BT_BREDR)
-	if (conn->type == BT_CONN_TYPE_BR || !bt_dev.le.acl_mtu) {
-		return bt_dev.br.mtu;
+	/* Check if the controller can accept ACL packets */
+	if (k_sem_take(bt_conn_get_pkts(conn), K_NO_WAIT)) {
+		/* not `goto fail`, we don't want to free the tx context: in the
+		 * case where it is the original buffer, it will contain the
+		 * callback ptr.
+		 */
```
[the remark this answers is not in the document]
Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c:613]: indeed it was an outdated comment, I was referring to `fail` in `do_send_frag`.

**Comment by prathje** [comment]
> #37577 was very useful for reproducing the deadlock in simulation, so thanks a lot @prathje ! You might want to review this as well, or try it on hardware.

Nice, glad to hear! Sadly, I am currently not working with BLE and unable to review this in detail. Thanks for mentioning me though, I am pleased to see the progress :)

File: subsys/bluetooth/host/conn.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/host/conn.c]: Even if they are just printed, it would still be a massive amount of data being printed :)

**Review by Thalley — APPROVED** [review]

File: subsys/bluetooth/host/conn.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (jori-nordic) [reply subsys/bluetooth/host/conn.c]: yep that is true. especially since I was testing with that l2cap stress test :p

File: tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/src/main.c [hunk 12]
```diff
@@ -372,6 +386,11 @@ static void test_central_main(void)
 	LOG_DBG("*L2CAP STRESS Central started*");
 	int err;
 
+	/* Prepare tx_data */
+	for (size_t i = 0; i < sizeof(tx_data); i++) {
+		tx_data[i] = (uint8_t)i;
```
Comment (jhedberg) [inline tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/src/main.c:391]: The type of the variable being assigned to (`tx_data[i]`) is already uint8_t, so truncation happens either way, ie the explicit typecast is redundant.

Comment (jori-nordic) [reply tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/src/main.c:391]: at least with the catch it's visible that it's on purpose and not a mistake

**Review by jhedberg — APPROVED** [review]
